src/lib/comms/include/sol-gatt.h

```diff
@@ -0,0 +1,256 @@
+ */
+
+enum sol_gatt_chr_flags {
+	SOL_GATT_CHR_FLAGS_BROADCAST = (1 << 0),
+	SOL_GATT_CHR_FLAGS_READ = (1 << 1),
+	SOL_GATT_CHR_FLAGS_WRITE_WITHOUT_RESPONSE = (1 << 2),
+	SOL_GATT_CHR_FLAGS_WRITE = (1 << 3),
+	SOL_GATT_CHR_FLAGS_NOTIFY = (1 << 4),
+	SOL_GATT_CHR_FLAGS_INDICATE = (1 << 5),
+	SOL_GATT_CHR_FLAGS_AUTHENTICATED_SIGNED_WRITES = (1 << 6),
+	SOL_GATT_CHR_FLAGS_RELIABLE_WRITE = (1 << 7),
+	SOL_GATT_CHR_FLAGS_WRITABLE_AUXILIARIES = (1 << 8),
+	SOL_GATT_CHR_FLAGS_ENCRYPT_READ = (1 << 9),
+	SOL_GATT_CHR_FLAGS_ENCRYPT_WRITE = (1 << 10),
+	SOL_GATT_CHR_FLAGS_ENCRYPT_AUTHENTICATED_READ = (1 << 11),
+	SOL_GATT_CHR_FLAGS_ENCRYPT_AUTHENTICATED_WRITE = (1 << 12),
+};
+
+enum sol_gatt_desc_flags {
+	SOL_GATT_CHR_FLAGS_READ = (1 << 0),
+	SOL_GATT_CHR_FLAGS_WRITE = (1 << 1),
+	SOL_GATT_CHR_FLAGS_ENCRYPT_WRITE = (1 << 2),
+	SOL_GATT_CHR_FLAGS_ENCRYPT_AUTHENTICATED_READ = (1 << 3),
+	SOL_GATT_CHR_FLAGS_ENCRYPT_AUTHENTICATED_WRITE = (1 << 4),
+};
+
+/*
+ * FIXME: need to add errors that make sense for the BlueZ D-Bus API
+ * and "raw" GATT.
+ */
+enum sol_gatt_error {
+	SOL_GATT_NO_ERROR,
+	SOL_GATT_FAIL
+};
+
+enum sol_gatt_attr_type {
+	SOL_GATT_ATTR_TYPE_SERVICE,
+	SOL_GATT_ATTR_TYPE_CHARACTERISTIC,
+	SOL_GATT_ATTR_TYPE_DESCRIPTOR,
+};
+
+struct sol_gatt_pending;
+
+struct sol_gatt_attr *sol_gatt_pending_get_attr(const struct sol_gatt_pending *op);
+
+/**
+ * @brief Representation of a GATT Attribute
+ */
+struct sol_gatt_attr {
+	struct sol_bt_uuid uuid;
+	enum sol_gatt_attr_type type;
+	uint16_t flags;
+	/**
+	 * The response for this operation will be returned by calling
+	 * #sol_gatt_pending_reply() passing a valid buffer, or an error.
+	 */
+	int (*read)(struct sol_gatt_pending *op,
+		    uint16_t offset);
+	/**
+	 * The response for this operation will be returned by calling
+	 * #sol_gatt_pending_reply(), the @a buf parameter will be ignored.
+	 */
+	int (*write)(struct sol_gatt_pending *op,
+		     struct sol_buffer *buf,
+		     uint16_t offset);
+	/**
+	 * This operation doesn't require an reply.
+	 */
+	/*
+	 * FIXME: is there any case that it would make sense to
+	 * wait for a reply?
+	 */
+	int (*flush)(struct sol_gatt_pending *op,
+		     bool sync);
+	void *user_data;
+};
+
+/**
+ * @brief Returns a response to a async operation
+ *
+ * When a operation is performed in a attribute, the response to the operation
+ * is only returned when this function is called.
+ *
+ * @param pending Pending operation to be responded
+ * @param error Error to be returned
+ * @param buf Payload (can be NULL) to be returned
+ */
+int sol_gatt_pending_reply(struct sol_gatt_pending *pending,
+			   enum sol_gatt_error error,
+			   struct sol_buffer *buf);
+
+/*
+ * Helpers to the construction of the local attribute database
+ */
+#define SOL_GATT_SERVICE(service_uuid) {			\
+		.type = SOL_GATT_ATTR_TYPE_SERVICE,		\
+		.uuid = service_uuid,				\
+	}
+
+#define SOL_GATT_CHARACTERISTIC(uuid, flags, read, write,		\
```

Comment:
this and GATT_DESCRIPTOR macros will get out of hand to the reader. It's better to give then __VA_ARGS__ and then let the user call with: `SOL_GATT_CHARACTERISTIC(uuid, .flags = myflags, .read = myread...)`